platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/fingerprint/ConfigurationCacheFingerprintWriter.kt

```diff
@@ -204,6 +215,21 @@ class ConfigurationCacheFingerprintWriter(
         CompositeStoppable.stoppable(buildScopedWriter, projectScopedWriter).stop()
     }
 
+    override fun buildAdded(buildState: BuildState) {
+        captureBuildSrcPresence(buildState)
```

Comment:
I looked into doing it as late as possible, instead of from constructor + `buildAdded` callback - from `#close()`, for instance. That seems to work too. However, we lose the nice messages for when buildSrc is removed - instead we seem to get the same message we have in master. IOW, if we are ok with the less specific message, we could then only care about tracking the missing buildSrc case, to detect addition. For the present one, we already detected that (in a way that is similar to how we track actual changes to those files, I guess).  

When removing `build.gradle`:
```
Expected: "Calculating task graph as configuration cache cannot be reused because a buildSrc build at 'buildSrc' has been removed."
 but: was "Calculating task graph as configuration cache cannot be reused because an input to task ':buildSrc:jar' has changed."
```

When removing `settings.gradle`:
```
Expected: "Calculating task graph as configuration cache cannot be reused because a buildSrc build at 'buildSrc' has been removed."
 but: was "Calculating task graph as configuration cache cannot be reused because file 'buildSrc/settings.gradle' has been removed."
```

When removing `src/main/groovy/MyClass.groovy`:
```
Expected: "Calculating task graph as configuration cache cannot be reused because a buildSrc build at 'buildSrc' has been removed."
 but: was "Calculating task graph as configuration cache cannot be reused because an input to task ':buildSrc:compileGroovy' has changed."

```

Reply:
One alternative approach would be to detect only changing from absent to present, as we already detect changing from present to absent correctly.

@alllex If you agree, you can cherry pick this commit: https://github.com/gradle/gradle/commit/f80048d739d48962f9f70f792cdeff6806789db7